### ISSUE TITLE
Conditional effects

### ISSUE THREAD

**Issue by mikand** [aiplan4eu/upf#31]
### User Story
As a **TSB Developer** I want to express conditional effects to simplify the modeling of conditional behaviors.

A conditional effect is an action effect that takes place only when a condition is met. A normal effect states that `x := 5`, a conditional effect could be stated as `if phi then x := 5` (where `phi` is a condition expression) in an imperative programming language.

In the UPF, at the moment, we can only add normal effects (with the `add_effect()` method of the `Action` class). We need to extend this function to make it possible to write something like:
```python
act = Action(...)
...
act.add_effcet(x, 5) # x:=5, normal effect
act.add_effect(y, 10, condition=Or(a, b)) # if a or b then y := 10
```

### Acceptance Criteria

- [x] Conditional effects can be stated via code and stored in the `Action` class
- [x] Conditional effects can be printed in a suitable human-readable format when a problem or an action is printed
- [x] The PDDL printer is extended to support conditional effects
- [x] The problem kind is extended to detect when conditional effects are present

### Additional Material

- https://planning.wiki/ref/pddl/requirements#conditional-effects

### Attention Points

- We need to decide if we wan to extend the effects representation or if we want to add a new, special structure just for conditional effects within the action. I am in favor of the first option, but we need to think about this.

**Comment by alvalentini** [comment]
I like more the first option too. We can create a base class Effect and then the different kind of effects extend the base class.
We already have assignment and increase/decrease effects that can be represented as different effects.

**Comment by Framba-Luca** [comment]
I also like more the first option; and i like the idea of creating a new class Effect. 

We (me and @alvalentini) were talking about the possibility of making a new Effect Class:

`
class Effect():

fluent
value
condition
`
That represents a basic effect (where the condition can be None or True for unconditional effects).

(option 1) Then this class can be extended to add more information about the effect (like an else branch or some time constraint or things like that in general).

(option 2) Another option would also be to make just a class effect full of boolean values like is_conditional, but in my opinion this doesn't accept very well new type additions.

(option 3) Create 2 different class for basic assignment and conditional assignment. Here, since an assignment can be both conditional and something else (like an increment or a decrement), it would need to extend both class (but if I am not wrong, python support multy-class inheritance.

(option 4) The last option we came up with is the one where we create an abstract class Effect and then extend this one. This makes sense if can exists an effect that doesn't have an assignment.

Which option do you think is better? @mikand  @alvalentini

**Comment by alvalentini** [comment]
I think we can start with a class `Effect` that contains four attributes:
- `kind`, that for now can be ASSIGNMENT, INCREASE, DECREASE;
- `condition`, with default value True;
- `fluent_exp`;
- `value_exp`.

**Comment by mikand** [comment]
+1 for the `Effect` class. One important feature we need is then a getter in the action to "filter" for different kinds of effects (e.g. give me all the conditional effects, give me all the non-conditional effects, ...)

**Comment by mikand** [comment]
Fixed in #32
